### PR TITLE
Add protocols post

### PR DESCRIPTION
**Pull request by capitancambio**

### CODE REVIEW

File: _posts/2019-03-21-python-interfaces-a-la-go.markdown [hunk 1]
```diff
@@ -0,0 +1,153 @@
+---
+title: Python interfaces a la Golang
+layout: post
+category: news
+author: Javier Asensio-Cubero 
+banner: /assets/img/posts/2015-11-23-tech-jobs.jpg
+hex: 0e1720
+---
+
+For the last four years I have been jumping back and forth from Python and Go. The majority of the work I carry out as a Machine Learning Engineer at [Octopus Energy](https://octopus.energy/) is mostly focused on Python and the typical data science stack. We are working on medium/big scale data processing projects using all the usual suspects: keras, scikit-learn, pandas, airflow... namely I'm not doing any Go.
```
Comment (frederizia) [inline _posts/2019-03-21-python-interfaces-a-la-go.markdown]: back and forth *between* Python and Go

File: _posts/2019-03-21-python-interfaces-a-la-go.markdown [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+layout: post
+category: news
+author: Javier Asensio-Cubero 
+banner: /assets/img/posts/2015-11-23-tech-jobs.jpg
+hex: 0e1720
+---
+
+For the last four years I have been jumping back and forth from Python and Go. The majority of the work I carry out as a Machine Learning Engineer at [Octopus Energy](https://octopus.energy/) is mostly focused on Python and the typical data science stack. We are working on medium/big scale data processing projects using all the usual suspects: keras, scikit-learn, pandas, airflow... namely I'm not doing any Go. 
+
+There are two major things that I miss from Go when working on a Python project, static typing, and go interfaces. At octopus energy we try to enforce type hinting as much as possible in order to avoid silly mistakes (the ones I do more often) with great results, regardless its the intrinsic limitations.
```
Comment (frederizia) [inline _posts/2019-03-21-python-interfaces-a-la-go.markdown]: on a Python project: static typing, and go interfaces
I would use a colon

regardless *of* its intrinsic limitations

File: _posts/2019-03-21-python-interfaces-a-la-go.markdown [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+author: Javier Asensio-Cubero 
+banner: /assets/img/posts/2015-11-23-tech-jobs.jpg
+hex: 0e1720
+---
+
+For the last four years I have been jumping back and forth from Python and Go. The majority of the work I carry out as a Machine Learning Engineer at [Octopus Energy](https://octopus.energy/) is mostly focused on Python and the typical data science stack. We are working on medium/big scale data processing projects using all the usual suspects: keras, scikit-learn, pandas, airflow... namely I'm not doing any Go. 
+
+There are two major things that I miss from Go when working on a Python project, static typing, and go interfaces. At octopus energy we try to enforce type hinting as much as possible in order to avoid silly mistakes (the ones I do more often) with great results, regardless its the intrinsic limitations. 
+
+While dealing with data streams and specially with the `io` package in Python, it becomes a burden trying to keep a flexible interface while reusing elements of the standard library. Let's see the following example:
```
Comment (frederizia) [inline _posts/2019-03-21-python-interfaces-a-la-go.markdown]: *e*specially

File: _posts/2019-03-21-python-interfaces-a-la-go.markdown [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import pandas as pd
+
+
+def transform(input_file: str) -> pd.DataFrame:
+    df = pd.read_csv(input_file, index="index")
+    transformed_df = _transform(df)
+    return transformed_df
+```
+
+This implementation has some drawbacks, passing a string with a file path for pandas to load a csv file makes testing difficult as you would need to create an actual csv. Also, we restrict ourselves to local files for applying our data transform, you might want to load the stream from other sources, such as http connections or a S3 bucket.
```
Comment (frederizia) [inline _posts/2019-03-21-python-interfaces-a-la-go.markdown]: ...our data transform, *but/however/whilst* you might want to load...

File: _posts/2019-03-21-python-interfaces-a-la-go.markdown [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+import io
+
+
+def transform(input_io: io.TextIO) -> pd.DataFrame:
+    df = pd.read_csv(input_io, index="index")
+    transformed_df = _transform(df)
+    return transformed_df
+```
+
+This sounds sensible, you're using a class offered by the standard library and `pandas`will like it as it uses an intrinsic reading _protocol_ (more on this soon). this means that as long as what you send down `pd.read_csv` has method/function called `read`, pandas will be able to load the data. But the definition `io` package makes a burden to extend the behaviour of any of its classes. Many `ftp` or `S3` readers out there won't inherit from these classes as typing is not really big in the python ecosystem. Also, building _fake_ io subclasses to test is an ordeal as you need to create a bunch of empty stubs to extend them.
```
Comment (frederizia) [inline _posts/2019-03-21-python-interfaces-a-la-go.markdown]: and `pandas` will (missing a space)
(more on this soon). This  (not capitalised)

has *a* method/function called `read`

This sentence sounds odd but don't know how to rewrite right now:
But the definition `io` package makes a burden to extend the behaviour of any of its classes.

File: _posts/2019-03-21-python-interfaces-a-la-go.markdown [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+##  Python protocols
+
+Let's focus on the concept of _protocol_ in python. This not something new, you might know that calling `len()` in any object that contains the method `__len__` will just work. `__getitem__` , and the pickling protocol are more examples. Python, being loosely typed, will check programmatically that `__len__` exists before calling it in runtime. 
+
+If you've written any go code, this will sound familiar, but of course you would enforce your protocol into an actual interface. In go, there's no concept of classes or inheritance, just _structs_. Structs can have some functions attached to them called _methods_ (sorry the nomenclature makes the comparison a bit messy). You define interfaces as contracts that your struct needs to meet, just like protocols! But being strictly typed, this is enforced during compilation time. 
+
+Relying on type hinting to play along with protocols is known as [structural subtyping](https://www.python.org/dev/peps/pep-0544/) in python.
+
+You might be wondering how this is different from inheritance. Let's have a look to the following example where we will create a useful function to print the length of different objects.
```
Comment (frederizia) [inline _posts/2019-03-21-python-interfaces-a-la-go.markdown]: Let's have a look *at* the following example

File: _posts/2019-03-21-python-interfaces-a-la-go.markdown [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        pass
+
+
+def transform(input_reader: Reader) -> pd.DataFrame:
+    df = pd.read_csv(input_reader, index="index")
+    transformed_df = _transform(df)
+    return transformed_df
+
+```
+From this implementation we obtain function that accepts not only any _file\_like\_object_, such as files, `io.StringIO` or `io.BytesIO`, but also any other third party libraries to stream S3 buckets, ftp, sftp, http, as long as they are iterables and have a `read(i:int=-1)` method. No inheritance needed, no extra glue code to make this work with third party libraries or testing fakes.
```
Comment (frederizia) [inline _posts/2019-03-21-python-interfaces-a-la-go.markdown]: we obtain *a* function

**Comment by capitancambio** [comment]
@codeinthehole :raised_hands:
